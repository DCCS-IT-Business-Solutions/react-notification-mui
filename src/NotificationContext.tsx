import * as React from "react";
import {
  NotificationItem,
  NotificationType,
  INotificationItemProps,
  ICustomNotificationProps,
} from "./NotificationItem";
import * as moment from "moment";
import { SnackbarOrigin } from "@material-ui/core/Snackbar";
import { ThemeStyle } from "@material-ui/core/styles/createTypography";

export interface INotificationTypeProps {
  autoHideDuration?: number;
  enableClickAway: boolean;
  transitionDuration: number;
  snackBarStyle?: React.CSSProperties;
  snackBarMessageStyle?: React.CSSProperties;
  snackBarMessageVariant?: ThemeStyle;
  snackBarTitleStyle?: React.CSSProperties;
  snackBarTitleVariant?: ThemeStyle;
  action?: React.ReactNode;
}

export interface INotificationContextProps {
  displayOldestOnTop?: boolean;
  infoProps?: INotificationTypeProps;
  warningProps?: INotificationTypeProps;
  errorProps?: INotificationTypeProps;
  anchorOrigin?: SnackbarOrigin;
  children?: any;
}

const defaultProps = {
  anchorOrigin: { vertical: "top", horizontal: "right" } as SnackbarOrigin,
  infoProps: {
    transitionDuration: 750,
    enableClickAway: false,
    autoHideDuration: 2500,
    snackBarStyle: {
      backgroundColor: "green",
      color: "white",
      fontSize: 16,
    },
    snackBarMessageVariant: "body1",
    snackBarTitleVariant: "h6",
  } as INotificationTypeProps,
  warningProps: {
    transitionDuration: 750,
    enableClickAway: false,
    autoHideDuration: 2500,
    snackBarStyle: {
      backgroundColor: "orange",
      color: "white",
      fontSize: 16,
    },
    snackBarMessageVariant: "body1",
    snackBarTitleVariant: "h6",
  } as INotificationTypeProps,
  errorProps: {
    transitionDuration: 750,
    enableClickAway: false,
    snackBarStyle: {
      backgroundColor: "red",
      color: "white",
      fontSize: 16,
    },
    snackBarMessageVariant: "body1",
    snackBarTitleVariant: "h6",
  } as INotificationTypeProps,
};

export function NotificationContextProvider(tprops: INotificationContextProps) {
  const [notifications, setNotifications] = React.useState<
    INotificationItemProps[]
  >([]);

  const props = { ...defaultProps, ...tprops };

  function addInfo(message: string | React.ReactNode, title?: string) {
    const now = moment.now();
    const noti = {
      ...props.infoProps,
      id: now + Math.random(),
      timestamp: now,
      open: true,
      variant: NotificationType.info,
      remainingDuration: props.infoProps.autoHideDuration,
      message,
      title,
    };
    addNotification(noti);
  }

  function addWarning(message: string | React.ReactNode, title?: string) {
    const now = moment.now();
    const noti = {
      ...props.warningProps,
      id: now + Math.random(),
      timestamp: now,
      open: true,
      variant: NotificationType.warning,
      remainingDuration: props.infoProps.autoHideDuration,
      message,
      title,
    };
    addNotification(noti);
  }

  function addError(message: string | React.ReactNode, title?: string) {
    const now = moment.now();
    const noti = {
      ...props.errorProps,
      id: now + Math.random(),
      timestamp: now,
      open: true,
      variant: NotificationType.error,
      remainingDuration: props.infoProps.autoHideDuration,
      message,
      title,
    };
    addNotification(noti);
  }

  function addCustomNotification(notification: ICustomNotificationProps) {
    const now = moment.now();
    addNotification({
      transitionDuration: 750,
      enableClickAway: false,
      snackBarMessageVariant: "body1",
      snackBarTitleVariant: "h6",
      ...notification,
      id: now + Math.random(),
      remainingDuration: notification.autoHideDuration,
      timestamp: now,
      open: true,
      variant: NotificationType.custom,
    });
  }

  function addNotification(noti: INotificationItemProps) {
    if (props.displayOldestOnTop === true) {
      setNotifications([...notifications, { ...noti }]);
    } else {
      setNotifications([{ ...noti }, ...notifications]);
    }
  }

  function handleClose(id: number) {
    setNotifications(
      notifications.map((noti) => {
        return noti.id === id ? { ...noti, open: false } : { ...noti };
      })
    );
  }

  function handleExited(id: number) {
    setNotifications(
      notifications.filter((noti) => {
        if (noti.id === id) {
          return false;
        }
        return true;
      })
    );
  }

  function handleSetHeight(id: number, height: number) {
    setNotifications(
      notifications.map((noti) =>
        noti.id === id ? { ...noti, height } : { ...noti }
      )
    );
  }

  function handleCalcMarginTop(index: number) {
    let marginTop = 0;
    notifications
      .slice(0, index)
      .forEach((noti) => (marginTop += noti.height || 0));
    return marginTop + index * 8;
  }

  function calcRemainingDuration(noti: INotificationItemProps) {
    if (noti.autoHideDuration) {
      return noti.autoHideDuration - (moment.now() - noti.timestamp);
    }
    return undefined;
  }

  return (
    <NotificationContext.Provider
      value={
        {
          addInfo,
          addWarning,
          addError,
          addCustomNotification,
        } as INotificationContext
      }
    >
      {props.children}
      {notifications.map((noti: INotificationItemProps, index: number) => (
        <NotificationItem
          {...noti}
          key={noti.id}
          index={index}
          onClose={handleClose}
          onExited={handleExited}
          remainingDuration={calcRemainingDuration(noti)}
          onSetHeight={handleSetHeight}
          onCalcMarginTop={handleCalcMarginTop}
          anchorOrigin={props.anchorOrigin}
        />
      ))}
    </NotificationContext.Provider>
  );
}

export interface INotificationContext {
  addInfo: (
    message: string | React.ReactNode,
    title?: string | React.ReactNode
  ) => void;
  addWarning: (
    message: string | React.ReactNode,
    title?: string | React.ReactNode
  ) => void;
  addError: (
    message: string | React.ReactNode,
    title?: string | React.ReactNode
  ) => void;
  addCustomNotification: (notification: ICustomNotificationProps) => void;
}

export const NotificationContext = React.createContext<
  INotificationContext | any
>({});
