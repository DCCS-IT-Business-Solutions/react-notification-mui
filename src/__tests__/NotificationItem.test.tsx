import * as React from "react";
import { shallow, mount, render } from "enzyme";
import { NotificationContextProvider } from "../NotificationContext";
import { Typography, Button } from "@material-ui/core";
import { NotificationItem, NotificationType } from "../NotificationItem";

it("should render", () => {
  const sut = render(
    <NotificationItem
      anchorOrigin={{ horizontal: "center", vertical: "bottom" }}
      id={1}
      index={0}
      message="Test"
      onCalcMarginTop={() => 10}
      onClose={() => {
        return;
      }}
      onExited={() => {
        return;
      }}
      open={true}
      enableClickAway={false}
      timestamp={150}
      variant={NotificationType.custom}
      transitionDuration={750}
    />
  );
  expect(sut).toMatchSnapshot();
});

it("should render with all options", () => {
  const sut = render(
    <NotificationItem
      anchorOrigin={{ horizontal: "center", vertical: "bottom" }}
      id={1}
      index={0}
      message="Test"
      title="Test"
      onCalcMarginTop={() => 10}
      onClose={() => {
        return;
      }}
      onExited={() => {
        return;
      }}
      open={true}
      enableClickAway={false}
      timestamp={150}
      variant={NotificationType.custom}
      transitionDuration={750}
      action={<Button>Test</Button>}
      autoHideDuration={700000}
      remainingDuration={700000}
      snackBarMessageStyle={{ fontSize: "100px", color: "pink" }}
      snackBarTitleStyle={{ fontSize: "100px", color: "red" }}
      snackBarMessageVariant="h1"
      snackBarTitleVariant="h5"
      children={<div>Test</div>}
    />
  );
  expect(sut).toMatchSnapshot();
});
