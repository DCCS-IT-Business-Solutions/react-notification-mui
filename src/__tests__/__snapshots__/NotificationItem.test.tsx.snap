// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter",
      "style": "margin-top:10px;z-index:10003",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiTypography-root MuiPaper-root MuiPaper-elevation6 MuiSnackbarContent-root MuiTypography-body2",
          "direction": "up",
          "role": "alertdialog",
          "style": "opacity:0;transform:scale(0.75, 0.5625)",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiSnackbarContent-message",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "MuiTypography-root MuiTypography-body1",
                    },
                    "children": Array [
                      Object {
                        "data": "Test",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "MuiSnackbarContent-action",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "MuiButtonBase-root MuiIconButton-root",
                    "tabindex": "0",
                    "type": "button",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "MuiIconButton-label",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "aria-hidden": "true",
                            "class": "MuiSvgIcon-root",
                            "focusable": "false",
                            "role": "presentation",
                            "style": "color:#ffffff",
                            "viewBox": "0 0 24 24",
                          },
                          "children": Array [
                            Object {
                              "attribs": Object {
                                "d": "M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z",
                              },
                              "children": Array [],
                              "name": "path",
                              "namespace": "http://www.w3.org/2000/svg",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {
                                "d": undefined,
                              },
                              "x-attribsPrefix": Object {
                                "d": undefined,
                              },
                            },
                          ],
                          "name": "svg",
                          "namespace": "http://www.w3.org/2000/svg",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "aria-hidden": undefined,
                            "class": undefined,
                            "focusable": undefined,
                            "role": undefined,
                            "style": undefined,
                            "viewBox": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "aria-hidden": undefined,
                            "class": undefined,
                            "focusable": undefined,
                            "role": undefined,
                            "style": undefined,
                            "viewBox": undefined,
                          },
                        },
                      ],
                      "name": "span",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "button",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "tabindex": undefined,
                    "type": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "tabindex": undefined,
                    "type": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "MuiSnackbarContent-action",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "MuiButtonBase-root MuiIconButton-root",
                  "tabindex": "0",
                  "type": "button",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "MuiIconButton-label",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "aria-hidden": "true",
                          "class": "MuiSvgIcon-root",
                          "focusable": "false",
                          "role": "presentation",
                          "style": "color:#ffffff",
                          "viewBox": "0 0 24 24",
                        },
                        "children": Array [
                          Object {
                            "attribs": Object {
                              "d": "M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z",
                            },
                            "children": Array [],
                            "name": "path",
                            "namespace": "http://www.w3.org/2000/svg",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {
                              "d": undefined,
                            },
                            "x-attribsPrefix": Object {
                              "d": undefined,
                            },
                          },
                        ],
                        "name": "svg",
                        "namespace": "http://www.w3.org/2000/svg",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "aria-hidden": undefined,
                          "class": undefined,
                          "focusable": undefined,
                          "role": undefined,
                          "style": undefined,
                          "viewBox": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "aria-hidden": undefined,
                          "class": undefined,
                          "focusable": undefined,
                          "role": undefined,
                          "style": undefined,
                          "viewBox": undefined,
                        },
                      },
                    ],
                    "name": "span",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "tabindex": undefined,
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "tabindex": undefined,
                  "type": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "MuiSnackbarContent-message",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "MuiTypography-root MuiTypography-body1",
                      },
                      "children": Array [
                        Object {
                          "data": "Test",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "p",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "direction": undefined,
          "role": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "direction": undefined,
          "role": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`should render with all options 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter",
      "style": "margin-top:10px;z-index:10003",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiTypography-root MuiPaper-root MuiPaper-elevation6 MuiSnackbarContent-root MuiTypography-body2",
          "direction": "up",
          "role": "alertdialog",
          "style": "opacity:0;transform:scale(0.75, 0.5625)",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiSnackbarContent-message",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "class": "MuiTypography-root MuiTypography-h5",
                          "style": "font-size:100px;color:red",
                        },
                        "children": Array [
                          Object {
                            "data": "Test",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "text",
                          },
                        ],
                        "name": "h5",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                          "style": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                          "style": undefined,
                        },
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {},
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "MuiTypography-root MuiTypography-h1",
                            "style": "font-size:100px;color:pink",
                          },
                          "children": Array [
                            Object {
                              "data": "Test",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                          ],
                          "name": "h1",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  Object {
                    "attribs": Object {},
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "class": "MuiTypography-root MuiTypography-h1",
                          "style": "font-size:100px;color:pink",
                        },
                        "children": Array [
                          Object {
                            "data": "Test",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "text",
                          },
                        ],
                        "name": "h1",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                          "style": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                          "style": undefined,
                        },
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {},
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "MuiTypography-root MuiTypography-h5",
                            "style": "font-size:100px;color:red",
                          },
                          "children": Array [
                            Object {
                              "data": "Test",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                          ],
                          "name": "h5",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "MuiSnackbarContent-action",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "MuiButtonBase-root MuiButton-root MuiButton-text",
                    "tabindex": "0",
                    "type": "button",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "MuiButton-label",
                      },
                      "children": Array [
                        Object {
                          "data": "Test",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "span",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "button",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "tabindex": undefined,
                    "type": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "tabindex": undefined,
                    "type": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "MuiSnackbarContent-action",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "MuiButtonBase-root MuiButton-root MuiButton-text",
                  "tabindex": "0",
                  "type": "button",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "MuiButton-label",
                    },
                    "children": Array [
                      Object {
                        "data": "Test",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "span",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "tabindex": undefined,
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "tabindex": undefined,
                  "type": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "MuiSnackbarContent-message",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "attribs": Object {},
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "MuiTypography-root MuiTypography-h5",
                            "style": "font-size:100px;color:red",
                          },
                          "children": Array [
                            Object {
                              "data": "Test",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                          ],
                          "name": "h5",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": Object {
                        "attribs": Object {},
                        "children": Array [
                          Object {
                            "attribs": Object {
                              "class": "MuiTypography-root MuiTypography-h1",
                              "style": "font-size:100px;color:pink",
                            },
                            "children": Array [
                              Object {
                                "data": "Test",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "text",
                              },
                            ],
                            "name": "h1",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {
                              "class": undefined,
                              "style": undefined,
                            },
                            "x-attribsPrefix": Object {
                              "class": undefined,
                              "style": undefined,
                            },
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": [Circular],
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    Object {
                      "attribs": Object {},
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "MuiTypography-root MuiTypography-h1",
                            "style": "font-size:100px;color:pink",
                          },
                          "children": Array [
                            Object {
                              "data": "Test",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                          ],
                          "name": "h1",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                            "style": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": Object {
                        "attribs": Object {},
                        "children": Array [
                          Object {
                            "attribs": Object {
                              "class": "MuiTypography-root MuiTypography-h5",
                              "style": "font-size:100px;color:red",
                            },
                            "children": Array [
                              Object {
                                "data": "Test",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "text",
                              },
                            ],
                            "name": "h5",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {
                              "class": undefined,
                              "style": undefined,
                            },
                            "x-attribsPrefix": Object {
                              "class": undefined,
                              "style": undefined,
                            },
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": [Circular],
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "direction": undefined,
          "role": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "direction": undefined,
          "role": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
